insights/client/crypto.py

```diff
@@ -0,0 +1,206 @@
+import errno
+import os.path
+import logging
+import shutil
+import tempfile
+import subprocess
+
+logger = logging.getLogger(__name__)
+
+
+class GPGCommandResult(object):
+    """Output of a GPGCommand.
+
+    Attributes:
+        ok (bool): Result of an operation.
+        return_code (int): Return code of the operation.
+        stdout (str): Standard output of the command.
+        stderr (str): Standard error of the command.
+        _command (str | None):
+            An optional reference to the GPGCommand object that created the result
+    """
+    def __init__(self, ok, return_code, stdout, stderr, command):
+        self.ok = ok
+        self.return_code = return_code
+        self.stdout = stdout
+        self.stderr = stderr
+        self._command = command
+
+    def __str__(self):
+        return (
+            "<{cls} ok={ok} return_code={code} stdout={out} stderr={err}>"
+        ).format(
+            cls=self.__class__.__name__,
+            ok=self.ok, code=self.return_code,
+            out=self.stdout, err=self.stderr,
+        )
```

Comment:
Even though this is _only_ a data structure, there is still a lot of complexity.

Reply:
It's not that complex, it's just a format string ;)

Reply:
A namedtuple comes with `__str__` and `__repr__` providing almost exactly the same string.

```python
>>> from collections import namedtuple
>>> GPGCommandResult = namedtuple("GPGCommandResult", ["ok", "return_code", "stdout", "stderr", "command"])
>>> result = GPGCommandResult(True, 0, "out", "", "echo out")
>>> print "Result: %s" % (result,)
GPGCommandResult(True, 0, 'out', '', 'echo out')
```

Although the format string itself is not that complex, the complexity I was refering to is the fact that it is completely redundant. An 8-line method of a custom class inevitably increases the code complexity, even though is as simple as this one. If you encounter it, you still need to read it and understand what it does and why it is there.